quil-rs/src/program/scheduling/graph.rs

```diff
@@ -440,107 +429,48 @@ impl<'a> InstructionBlock<'a> {
             add_dependency!(graph, dependency.node_id => ScheduledGraphNode::BlockEnd, execution_dependency);
         }
 
-        Ok(InstructionBlock {
-            graph,
-            instructions,
-            terminator: terminator.unwrap_or(BlockTerminator::Continue),
-        })
+        Ok(ScheduledBasicBlock { graph, basic_block })
     }
 
     pub fn get_dependency_graph(&self) -> &DependencyGraph {
         &self.graph
     }
 
+    pub fn instructions(&'a self) -> &[&'a Instruction] {
+        self.basic_block.instructions()
+    }
+
     /// Return a particular-indexed instruction (if present).
     pub fn get_instruction(&self, node_id: usize) -> Option<&Instruction> {
-        self.instructions.get(node_id).copied()
+        self.instructions().get(node_id).copied()
+    }
+
+    pub fn label(&self) -> Option<&Target> {
+        self.basic_block.label()
     }
 
     /// Return the count of executable instructions in this block.
     pub fn len(&self) -> usize {
-        self.instructions.len()
+        self.instructions().len()
     }
 
     /// Return true if this block contains no executable instructions.
     pub fn is_empty(&self) -> bool {
-        self.instructions.is_empty()
+        self.instructions().is_empty()
     }
 
-    pub fn set_exit_condition(&mut self, terminator: BlockTerminator<'a>) {
-        self.terminator = terminator;
+    pub fn terminator(&self) -> &BasicBlockTerminator {
+        self.basic_block.terminator()
     }
-}
 
-#[derive(Clone, Debug, PartialEq)]
-pub enum BlockTerminator<'a> {
-    Conditional {
-        condition: &'a MemoryReference,
-        target: &'a str,
-        jump_if_condition_true: bool,
-    },
-    Unconditional {
-        target: &'a str,
-    },
-    Continue,
-    Halt,
+    pub fn basic_block(&self) -> &BasicBlock<'a> {
+        &self.basic_block
+    }
 }
 
 #[derive(Clone, Debug)]
 pub struct ScheduledProgram<'a> {
-    /// All blocks within the ScheduledProgram, keyed on string label.
```

Comment:
This logic was split up into two steps:

* build CFG
* schedule blocks

in the course of the change, it got simpler and more maintainable